### PR TITLE
An implementation of TFQMR

### PR DESCRIPTION
**Pull request by amontoison**
Version of TFQMR where symmetric Givens cosines/sines (c,s) are used instead of the classic formulation with tangents/cosines (θ,c). Like CGS, it's right preconditioning.

tfqmr1
tfqmr2

### CODE REVIEW

File: src/tfqmr.jl [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+  status = "unknown"
+
+  while !(solved || tired)
+
+    # σₘ = < vₘ,r₀ >
+    σ = @kdot(n, v, r₀)
+    # αₘ = ρₘ / σₘ
+    α = ρ / σ
+    
+    # w₂ₘ₊₁ = w₂ₘ - αₘ * u₂ₘ
```
Comment (dpo) [inline src/tfqmr.jl]: Same comment as for CGS. I think this part of the implementation would be more readable if these comments were to the right of the statements.

File: src/tfqmr.jl [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    @kscal!(n, β, v)
+    # z₂ₘ₊₂ = M⁻¹ * y₂ₘ₊₂
+    z = M * y
+    # u₂ₘ₊₂ = A * z₂ₘ₊₂
+    u = A * z
+    # vₘ₊₁ = u₂ₘ₊₂ + vₐᵤₓ
+    @kaxpy!(n, 1.0, u, v)
+    
+    # Update iteration index.
+    iter = iter + 1 # m ← m + 1
```
Comment (dpo) [inline src/tfqmr.jl]: Is this comment really necessary? :-)

File: test/test_tfqmr.jl [hunk 3]
```diff
@@ -0,0 +1,62 @@
+include("get_div_grad.jl")
```
Comment (dpo) [inline test/test_tfqmr.jl]: It's ok to test on symmetric systems, but it should also work on unsymmetric systems.

**Comment by dpo** [comment]
This branch needs to be rebased too.

**Comment by Xx-fractale-xX** [comment]
I rebased `cgs` branch. I wait that you merge it before I rebase this one. Otherwise new conflits will be generated.

File: src/tfqmr.jl [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+#
+# Alexis Montoison, <alexis.montoison@polymtl.ca>
+# Montreal, October 2018.
+
+export tfqmr
+
+"""Solve the consistent linear system Ax = b using the transpose-free quasi-minimal residual algorithm.
+TFMQR may be used to solve unsymmetric systems.
+
+This implementation allows a right preconditioning with M.
```
Comment (dpo) [inline src/tfqmr.jl]: ... allows a right preconditioner M.

**Comment by dpo** [comment]
The implementation doesn't quite follow the pseudo-code. Could you post the pseudo-code that you used as a guide?

**Comment by Xx-fractale-xX** [comment]
It's the pseudo-code that I used as guide but I did some modifications.
I computed `y₂` and `u₂` at the end of the iteration instead at the beginning in `(b)` step.
I did that because all variables were initialized before. 
In the code I used the same `y` for `y₁` and `y₂`, I just say that even coefficients (₂ₘ) represent `y₁`, `u₁` and odd coefficients (₂ₘ₊₁) represent `y₂`, `u₂`. It's the same thing for the other variables.
The last modification I did is to used "LAPACK" version of Givens reflection instead of the basic Givens rotation formulation. And  for that I needed to write  `θ`  as `s / c`.

**Comment by dpo** [comment]
This will need to be adjusted for Julia 0.7.

**Comment by Xx-fractale-xX** [comment]
I adjusted it and add `BLAS.axpby!` function which was added with julia 0.7.

**Comment by Xx-fractale-xX** [comment]
Done. I also add an allocation test for `DIOM` and `DQGMRES`. I commented it the time that https://github.com/JuliaArrays/ArrayViews.jl/pull/66 is merged. We will be able to `view` without allocations.

**Comment by dpo** [comment]
Oops, the PRs need to be rebased for some reason.

**Comment by Xx-fractale-xX** [comment]
I recommand to merge this PR at the end. I had `:tfqmr` in `test_variants.jl`. It will break the other PRs.

**Comment by dpo** [comment]
Seems like an epidemic: 
https://github.com/JuliaCI/Coverage.jl/issues/187

**Comment by Xx-fractale-xX** [comment]
I rebased this branch, I think we could merge it now. In all case, we need to rebase each time that we add an allocation test...

**Comment by dpo** [comment]
Sorry, this needs to be rebased again.

**Comment by Xx-fractale-xX** [comment]
Done.
Just a little comment, I remember why I did right preconditioning (`DIOM`, `DQGMRES`, `CGS`, `TFQMR`). In all those methods it's not necessary to resolve `M^{-1}x = y` after, it's already done in the methods. But the huge advantage is that the residual norm of the problem is not modified. With a left preconditioning, each iteration we have the residual norm of the modifed system `M^{-1}(Ax_m -b)` whereas we have the residual norm of the initial system that we want to solve with right preconditioning.

File: src/tfqmr.jl [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    ρ_next = @kdot(n, w, b)                       # ρₘ₊₁ = < w₂ₘ₊₂,r₀ >
+    β = ρ_next / ρ                                # βₘ = ρₘ₊₁ / ρₘ
+    ρ = ρ_next                                    # ρₘ ← ρₘ₊₁
+
+    @kaxpby!(n, 1.0, w, β, y)                     # y₂ₘ₊₂ = w₂ₘ₊₂ + βₘ * y₂ₘ₊₁
+    @kaxpby!(n, 1.0, u, β, v); @kscal!(n, β, v)   # vₐᵤₓ = βₘ * (u₂ₘ₊₁ + βₘ * vₘ)
+    z = M * y                                     # z₂ₘ₊₂ = M⁻¹ * y₂ₘ₊₂
+    u = A * z                                     # u₂ₘ₊₂ = A * z₂ₘ₊₂
+
+    @kaxpy!(n, 1.0, u, v)                         # vₘ₊₁ = u₂ₘ₊₂ + vₐᵤₓ
```
Comment (dpo) [inline src/tfqmr.jl]: It's still a mystery to me why certain utf-8 characters don't display properly. @abelsiqueira Do they all display properly for you?

Comment (abelsiqueira) [reply src/tfqmr.jl]: On my computer, both from the e-mail and GitHub they display properly. But from my cell, they don't.

Comment (dpo) [reply src/tfqmr.jl]: Some linuxy thing is going on...

Comment (abelsiqueira) [reply src/tfqmr.jl]: Maybe a browser thing?

Comment (dpo) [reply src/tfqmr.jl]: No because if I download the file locally, those symbols don't display properly in the editor either.

Comment (Xx-fractale-xX) [reply src/tfqmr.jl]: I use the module `UnicodeMath` for SublimeText. We can see all unicode character available there: https://github.com/mvoidex/UnicodeMath/blob/master/table.md

Comment (dpo) [reply src/tfqmr.jl]: What font do you use in Sublime Text? I downloaded and installed Lucida Sans Unicode, and still, the characters don't display correctly.

Could you open tfqmr.jl in Atom and see if the characters display properly?

Comment (Xx-fractale-xX) [reply src/tfqmr.jl]: I use the `Monospace` font. It's the default one with Sublime Text (linux version).

I opened `tfqmr.jl` in Atom. I installed nothing and I can still see characters.
capture d ecran de 2018-12-26 21-56-58

**Comment by amontoison** [comment]
Rebased. `TFQMR` use by default `opEye()` now and can resolve systems in `Float16` and `BigFloat` precision.

**Pull request closed without merging**